src/System.IO.FileSystem/src/System/IO/Directory.cs

```diff
@@ -268,22 +268,33 @@ public static void Move(string sourceDirName, string destDirName)
             string fulldestDirName = Path.GetFullPath(destDirName);
             string destPath = PathInternal.EnsureTrailingSeparator(fulldestDirName);
 
-            StringComparison pathComparison = PathInternal.StringComparison;
+            bool sameDirectoryIgnoreCase =
+                string.Equals(sourcePath, destPath, StringComparison.OrdinalIgnoreCase);
 
-            if (string.Equals(sourcePath, destPath, pathComparison))
+            // If the paths are the exact same, fail.
+            if (sameDirectoryIgnoreCase
+                && string.Equals(sourcePath, destPath, StringComparison.Ordinal))
                 throw new IOException(SR.IO_SourceDestMustBeDifferent);
 
             string sourceRoot = Path.GetPathRoot(sourcePath);
             string destinationRoot = Path.GetPathRoot(destPath);
-            if (!string.Equals(sourceRoot, destinationRoot, pathComparison))
+
+            // Compare paths for the same, skip this step if we already know the paths are identical.
+            if (!sameDirectoryIgnoreCase &&
+                !string.Equals(sourceRoot, destinationRoot, StringComparison.OrdinalIgnoreCase))
                 throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
 
             // Windows will throw if the source file/directory doesn't exist, we preemptively check
             // to make sure our cross platform behavior matches NetFX behavior.
             if (!FileSystem.DirectoryExists(fullsourceDirName) && !FileSystem.FileExists(fullsourceDirName))
                 throw new DirectoryNotFoundException(SR.Format(SR.IO_PathNotFound_Path, fullsourceDirName));
 
-            if (FileSystem.DirectoryExists(fulldestDirName))
+            if (!sameDirectoryIgnoreCase // This check is to allowing renaming of directories
+                && FileSystem.DirectoryExists(fulldestDirName))
+                throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
+
+            // If the direcotires aren't the same and the OS says the directory exists already, fail.
+            if (!sameDirectoryIgnoreCase && Directory.Exists(fulldestDirName))
                 throw new IOException(SR.Format(SR.IO_AlreadyExists_Name, fulldestDirName));
 
             FileSystem.MoveDirectory(fullsourceDirName, fulldestDirName);
```

Comment:
Let's assume you have a folder "C:\FOO\BAR"

What would the result be is if you call the method with these changes, using the following two arguments?:

> `sourceDirName = "C:\\foo\\BAR"`
> `destDirName = "C:\\FOO\\BAR"`

From what I understand:

1. Line 271: `sameDirectoryIgnoreCase` is `true`. That's ok.
2. Line 275: `sameDirectoryIgnoreCase` is `true` but the ordinal comparison is `false`, so we won't throw here. Should we have thrown at this point? The leaf directory was the exact same, but the parent was not.
3. Line 283: `!sameDirectoryIgnoreCase` evaluates to `false`, so we will not do the second check where we should verify if the roots are in the same volume. This verification should not be skipped because Windows does not allow moving between volumes.
4. Line 289: The source FileSystem directory exists and the source FileSystem file exists, so we won't throw. That's ok.
5. Line 292: Again, `!sameDirectoryIgnoreCase` evaluates to `false`, so we will not do the second check where we should verify if the destination FileSystem directory exists, even though we should.
6. Line 297: `!sameDirectoryIgnoreCase` evaluates to `false`, so we don't check if the destination Directory exists, even though we should.
7. Steps 5 and 6 made me wonder if we should be evaluating both `FileSystem.Exists` and `Directory.Exists` in the same `if` instead of separately.

We would end up reaching line 300, and I am not sure what the expected result would be.

cc @JeremyKuhne

Reply:
@carlossanlop as @JeremyKuhne  hasn't responded. I should check if the paths are ordinally (or whatever the OS says) the same, other than directory names, if the paths are the same, but the names are case-variants of themselves. I should continue, otherwise fail for question 2.

Do you agree? This PR is kind of in limbo now and I'd like to get it closed out

Reply:
Sorry @Jlalond! Been distracted with WinForms work. Your answer seems plausible to me. @carlossanlop can you sanity check this?

I had a test question that I've responded to as well.

Reply:
The changes are more clear now. Thanks for addressing. I see you also added a unit test to cover this example.